packages/go_router_builder/lib/src/type_helpers.dart

```diff
@@ -42,20 +44,44 @@ const List<_TypeHelper> _helpers = <_TypeHelper>[
   _TypeHelperString(),
   _TypeHelperUri(),
   _TypeHelperIterable(),
+  _TypeHelperJson(),
 ];
 
 /// Returns the decoded [String] value for [element], if its type is supported.
 ///
 /// Otherwise, throws an [InvalidGenerationSourceError].
-String decodeParameter(ParameterElement element, Set<String> pathParameters) {
+String decodeParameter(ParameterElement element, Set<String> pathParameters,
+    List<ElementAnnotation>? metadata) {
   if (element.isExtraField) {
     return 'state.${_stateValueAccess(element, pathParameters)}';
   }
 
   final DartType paramType = element.type;
   for (final _TypeHelper helper in _helpers) {
     if (helper._matchesType(paramType)) {
-      String decoded = helper._decode(element, pathParameters);
+      String? decoder;
+
+      final ElementAnnotation? annotation =
+          metadata?.firstWhereOrNull((ElementAnnotation annotation) {
+        return annotation.computeConstantValue()?.type?.getDisplayString() ==
+            'CustomParameterCodec';
+      });
+      if (annotation != null) {
+        final String? decode = annotation
+            .computeConstantValue()
+            ?.getField('decode')
+            ?.toFunctionValue()
+            ?.displayName;
+        final String? encode = annotation
+            .computeConstantValue()
+            ?.getField('encode')
+            ?.toFunctionValue()
+            ?.displayName;
+        if (decode != null && encode != null) {
+          decoder = decode;
+        }
+      }
```

Comment:
nit: code here is same as L110-L129, you can extract them into a function to reuse.